### ISSUE TITLE
Validation unit tests + new validation endpoints

### ISSUE THREAD

**Issue by nleroy917** [pepkit/pephub#185]
## Overview
To better support server-side validation of PEPs, I think having unit tests for all validation scenarios would be really useful. This also drives the development of new validation endpoints. Validation requires two things: **A PEP and a schema**. Both of these components can be in one of three forms: 1) a registry path in our database (`namespace/name`), 2) a file, and 3) a raw string. We should be able to accept any combination of these. As such, there will be $3 \times 3=9$ possible validation "flows". We can construct a matrix representing these combinations like so:

<img width="500" alt="image" src="https://github.com/pepkit/pephub/assets/41063083/6782b451-b52c-4c98-ac0d-9649316f2e46">

We need an endpoint (or endpoints) that can validate a PEP and a schema for each box. For each box we need two unit tests: 1) one that passes validation, and 2) one that fails validation. As such, there will be **18 unit tests.**

## A note on FastAPI and content types
A [Content-Type header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) in an HTTP request is used to indicate the original media type of the resource. FastAPI has specific ways of dealing with specific Content-Types. In brief, there are two content types that we use in our server: 1) `application/json` and 2) `multipart/form-data`. I'll explain them in the context of FastAPI in turn:

### `application/json`
Overwhelmingly, we use the `application/json` content type for pephub. It is widely supported by HTTP clients and servers, making it a standard choice for exchanging structured data in JSON format over the web. FastAPI uses [the request body](https://fastapi.tiangolo.com/tutorial/body/) to capture `JSON` sent to us via `POST/PATCH/DELETE` requests and the `application/json` content-type. Here is an example:

```python
class Item(BaseModel):
    name: str
    description: Union[str, None] = None
    price: float
    tax: Union[float, None] = None


app = FastAPI()


@app.post("/items/")
async def create_item(item: Item, db=Depends(get_db)):
    db.create(user)
    return item
```

FastAPI abstracts away everything for us, and we simply provide a `pydantic` model where it 1) validates the data sent via the request, and 2) stuffs in the parameters so we can then use them in our server logic. That's it!

### `multipart/form-data`
In rare cases, we use the `multipart/form-data` content-type when we want to work with files on pephub. Like `application/json`, [FastAPI has a special way](https://fastapi.tiangolo.com/tutorial/request-forms/) we declare the parameters we expect to receive in a `multipart/form-data` request. In short, we use the `Form` object to declare that the route expects form values.

```python
from fastapi import FastAPI, Form

app = FastAPI()


@app.post("/login/")
async def login(username: str = Form(), password: str = Form()):
    return {"username": username}
```

Now we can submit requests to this endpoint using form data, and FastAPI will parse and validate everything for us, stuffing variables in the proper function params.

## The validation endpoints
The endpoints should (at least try) to use a combination of the above to create a flow that accepts anything. Two conclusions were made: 1) it might be possible to restrict the endpoint to `multipart/form-data` so that file submission is simple while maintaining the ability to send strings (paths and raw PEPs/schema); and 2) all fields should be made optional so everything is flexible and we can render logic that responds with error codes if things are incorrect.

## Conclusion
In conclusion, we need 1) a new validation endpoint (or maybe endpoints if thats easier) that accepts files, strings, and registry paths, and 2) unit tests to test them and provide examples. One complete, the PEPhub UI can be updated to support validation that actually works 😉

**Comment by nleroy917** [comment]
### Discussed June 27th, 2023
- It seems possible that we can create an endpoint that accept both `application/json` and `multipart-form`. See [this StackOverflow](https://stackoverflow.com/questions/74009210/how-to-create-a-fastapi-endpoint-that-can-accept-either-form-or-json-body).

**Comment by ayobi** [comment]
This is now in release `0.9.x`
